### PR TITLE
[W5.11][T16-4] Lim Yi Sheng

### PR DESCRIPTION
**Pull request by LimYiSheng**
- Added sort command.
- Updated user guide to include sort command.
- Updated tests to test for the newly implemented sort command

### CODE REVIEW

**Comment by ChuaZhenWei** [comment]
very good

**Comment by zhihong8888** [comment]
very good

**Comment by XiiaoPanda** [comment]
Very good!!

**Comment by ryanchen2018** [comment]
Very good!

File: src/seedu/addressbook/commands/Command.java [hunk 1]
```diff
@@ -36,6 +36,10 @@ public static String getMessageForPersonListShownSummary(List<? extends ReadOnly
         return String.format(Messages.MESSAGE_PERSONS_LISTED_OVERVIEW, personsDisplayed.size());
     }
 
+    public static String getMessageForSortedPersonListShownSummary(List<? extends ReadOnlyPerson> personDisplayed) {
```
Comment (ChuaZhenWei) [inline src/seedu/addressbook/commands/Command.java:39]: proper capitalizing of alphabet for better and clearer reading for users

File: src/seedu/addressbook/data/person/UniquePersonList.java [hunk 2]
```diff
@@ -81,6 +76,17 @@ public UniquePersonList(UniquePersonList source) {
         return Collections.unmodifiableList(internalList);
     }
 
+    public List<ReadOnlyPerson> sortedListView() {
+        this.internalList.sort(new Comparator<Person>() {
+            @Override
+            public int compare(Person person1, Person person2) {
+                String name1 = person1.getName().toString();
```
Comment (ChuaZhenWei) [inline src/seedu/addressbook/data/person/UniquePersonList.java:83]: proper naming of variables as per java coding standard

File: src/seedu/addressbook/common/Messages.java [hunk 3]
```diff
@@ -11,6 +11,7 @@
     public static final String MESSAGE_INVALID_PERSON_DISPLAYED_INDEX = "The person index provided is invalid";
     public static final String MESSAGE_PERSON_NOT_IN_ADDRESSBOOK = "Person could not be found in address book";
     public static final String MESSAGE_PERSONS_LISTED_OVERVIEW = "%1$d persons listed!";
+    public static final String MESSAGE_PERSONS_SORTED_OVERVIEW = "%1$d persons sorted and listed!";
```
Comment (ChuaZhenWei) [inline src/seedu/addressbook/common/Messages.java:14]: Java coding standards: White space within a statement, are done in the with proper spacing between words and characters

File: docs/UserGuide.adoc [hunk 4]
```diff
@@ -72,6 +73,11 @@ Examples:
 Shows a list of all persons, along with their non-private details, in the address book. +
 Format: `list`
 
+== Sorting contacts in address book and listing all persons : 'sort'
```
Comment (jordancjq) [inline docs/UserGuide.adoc:76]: That's good that you have updated the User Guide (UG) to show the newly implemented function that you have created. The UG is the place where the user will go to find out the different features and the usage of it most of the time.

File: src/seedu/addressbook/commands/Command.java (hunk 1 above)
Comment (jordancjq) [inline src/seedu/addressbook/commands/Command.java:39]: Would be good to have the header comments before the function to indicate what the function will take in as a parameter, similarly to the one before this newly created function.

File: src/seedu/addressbook/commands/HelpCommand.java [hunk 5]
```diff
@@ -21,6 +21,7 @@ public CommandResult execute() {
                 + "\n" + ClearCommand.MESSAGE_USAGE
                 + "\n" + FindCommand.MESSAGE_USAGE
                 + "\n" + ListCommand.MESSAGE_USAGE
+                + "\n" + SortCommand.MESSAGE_USAGE
```
Comment (jordancjq) [inline src/seedu/addressbook/commands/HelpCommand.java:24]: This is good, make sure to update your `help` command as well, so that internally it also reflects the new commands to the user.

File: src/seedu/addressbook/commands/SortCommand.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package seedu.addressbook.commands;
+
+import seedu.addressbook.data.person.ReadOnlyPerson;
+
+import java.util.List;
+
+/**
+ * Sorts contacts within AddressBook for the user
```
Comment (jordancjq) [inline src/seedu/addressbook/commands/SortCommand.java:8]: Good documentation before the class implementation to have a overview of what the class will be doing.

File: src/seedu/addressbook/data/person/Name.java [hunk 7]
```diff
@@ -59,5 +59,4 @@ public boolean equals(Object other) {
     public int hashCode() {
         return fullName.hashCode();
     }
-
```
Comment (jordancjq) [inline src/seedu/addressbook/data/person/Name.java:62]: A random deletion of a line from another file. Try to not make changes that are not related to the scope of your work, as it affects the readability of the PR.

File: src/seedu/addressbook/data/person/UniquePersonList.java [hunk 8]
```diff
@@ -1,11 +1,6 @@
 package seedu.addressbook.data.person;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.Iterator;
-import java.util.List;
+import java.util.*;
```
Comment (jordancjq) [inline src/seedu/addressbook/data/person/UniquePersonList.java:3]: This currently violates the java coding standards: **Imported classes should always be listed explicitly.**

You can prevent IntelliJ from doing so by going through,
`Settings/Preferences -> Code Style -> Java -> Check "Use single class import"`
and
`Set Class count to use import with '*' & Names count to use static import with '*' to > 100`

File: src/seedu/addressbook/data/person/UniquePersonList.java [hunk 9]
```diff
@@ -81,6 +76,17 @@ public UniquePersonList(UniquePersonList source) {
         return Collections.unmodifiableList(internalList);
     }
 
+    public List<ReadOnlyPerson> sortedListView() {
+        this.internalList.sort(new Comparator<Person>() {
```
Comment (jordancjq) [inline src/seedu/addressbook/data/person/UniquePersonList.java:80]: Great job in using the in-line comparator to do the sorting.

File: test/input.txt [hunk 10]
```diff
@@ -148,6 +148,18 @@
   clear
   list
 
+##########################################################
+# test sort command
+##########################################################
+  
+  # add some contacts into address book
+  add john p/12345678 e/hello@hotmail.com a/311, clementi ave 2, #02-25
```
Comment (jordancjq) [inline test/input.txt:156]: Good job in adding a test case for the IO, maybe you can start trying out the JUnit test suite to familiarise yourself with it.

**Review by jordancjq — APPROVED** [review]
Good attempt for the enhancement this week. Please check the review comments and close the PR. Remember to put header comments for function as well to improve the readability of the functions when needed.
